scripts/west_commands/sign.py

```diff
@@ -220,59 +234,118 @@ def sign(self, command, build_dir, bcfg, formats):
+        if in_hex:
             out_hex = args.shex or str(b / 'zephyr' / 'zephyr.signed.hex')
-            log.inf('Generating:', out_hex)
-            sign_hex = (sign_base + args.tool_args +
-                        [runner_config.hex_file, out_hex])
+            sign_hex = sign_base + [in_hex, out_hex]
+            log.inf('hex: {}'.format(out_hex))
             log.dbg(quote_sh_list(sign_hex))
             subprocess.check_call(sign_hex)
 
+    @staticmethod
+    def imgtool(command, args):
+        if args.tool_path:
+            command.check_force(shutil.which(args.tool_path),
+                                '--tool-path {}: not an executable'.
+                                format(args.tool_path))
+            tool_path = args.tool_path
+        else:
+            tool_path = shutil.which('imgtool')
+            if not tool_path:
+                log.die('imgtool not found; either install it',
+                        '(e.g. "pip3 install imgtool") or provide --tool-path')
+        return tool_path
+
     @staticmethod
     def get_cfg(command, bcfg, item):
         try:
             return bcfg[item]
         except KeyError:
             command.check_force(
-                False,
-                "imgtool parameter unknown, build directory has no {} {}".
-                format('device tree define' if item.startswith('DT_') else
-                       'Kconfig option',
-                       item))
+                False, "build .config is missing a {} value".format(item))
             return None
+
+    @staticmethod
+    def dt_params(b, cache):
+        # Get the flash device's write alignment and the image-0
+        # partition's size out of the build directory's devicetree.
+
+        # Retrieve the list of devicetree bindings from cache.
+        try:
+            bindings = cache.get_list('CACHED_DTS_ROOT_BINDINGS')
+            log.dbg('DTS bindings:', bindings, level=log.VERBOSE_VERY)
+        except KeyError:
+            log.die('CMake cache has no CACHED_DTS_ROOT_BINDINGS.'
+                    '\n  Try again after re-building your application.')
+
+        # Ensure the build directory has a compiled DTS file
+        # where we expect it to be.
+        dts = b / 'zephyr' / (cache['CACHED_BOARD'] + '.dts.pre.tmp')
+        if not dts.is_file():
+            log.die("can't find DTS; expected:", dts)
+        log.dbg('DTS file:', dts, level=log.VERBOSE_VERY)
+
+        # Parse the devicetree using bindings from cache.
+        try:
+            edt = edtlib.EDT(dts, bindings)
+        except edtlib.EDTError as e:
+            log.die("can't parse devicetree:", e)
+
+        # By convention, the zephyr,flash chosen node contains the
+        # partition information about the zephyr image to sign.
+        flash = edt.chosen_node('zephyr,flash')
+        if not flash:
+            log.die('devicetree has no chosen zephyr,flash node;',
+                    "can't infer flash write block or image-0 slot sizes")
+
```

Comment:
Maybe there could be a helper function to fetch the flash node (or just to create the `EDT` instance), to shorten `dt_params()` a bit.